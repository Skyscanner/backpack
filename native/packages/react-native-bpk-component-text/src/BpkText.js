/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2017 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Text,
  Platform,
  StyleSheet,
} from 'react-native';
import React from 'react';
import PropTypes from 'prop-types';

const IOS_TOKENS = require('bpk-tokens/tokens/ios/base.react.native.common.js');
const ANDROID_TOKENS = require('bpk-tokens/tokens/android/base.react.native.common.js');

const tokens = Platform.select({
  ios: () => IOS_TOKENS,
  android: () => ANDROID_TOKENS,
})();

const TEXT_STYLES = ['xs', 'sm', 'base', 'lg', 'xl', 'xxl'];

const styleForTextStyle = (textStyle) => {
  const camelCasedStyle = textStyle[0].toUpperCase() + textStyle.slice(1);
  const determineAndroidFontFamily = (fontWeight) => {
    switch (fontWeight) {
      case '500':
        return 'sans-serif-medium';
      case '400':
      default:
        return 'sans-serif';
    }
  };

  return {
    fontSize: tokens[`text${camelCasedStyle}FontSize`],
    lineHeight: tokens[`text${camelCasedStyle}LineHeight`],
    fontWeight: Platform.OS === 'ios' ? tokens[`text${camelCasedStyle}FontWeight`] : undefined,
    letterSpacing: Platform.OS === 'ios' ? tokens[`text${camelCasedStyle}LetterSpacing`] : undefined,
    color: tokens.colorGray900,
    fontFamily: Platform.OS === 'android' ?
      determineAndroidFontFamily(tokens[`text${camelCasedStyle}FontWeight`]) : undefined,
  };
};

const styles = StyleSheet.create({
  xs: {
    ...styleForTextStyle('xs'),
  },
  sm: {
    ...styleForTextStyle('sm'),
  },
  base: {
    ...styleForTextStyle('base'),
  },
  lg: {
    ...styleForTextStyle('lg'),
  },
  xl: {
    ...styleForTextStyle('xl'),
  },
  xxl: {
    ...styleForTextStyle('xxl'),
  },
});

const BpkText = (props) => {
  const { children, textStyle, style: innerStyle, ...rest } = props;
  const style = styles[textStyle];
  const finalStyles = [style];

  if (innerStyle) {
    finalStyles.push(innerStyle);
  }

  return (
    <Text style={finalStyles} {...rest}>{children}</Text>
  );
};

BpkText.propTypes = {
  children: PropTypes.node.isRequired,
  textStyle: PropTypes.oneOf(TEXT_STYLES),
  style: Text.propTypes.style,
};

BpkText.defaultProps = {
  textStyle: 'base',
  style: {},
};

export default BpkText;
export { TEXT_STYLES };
