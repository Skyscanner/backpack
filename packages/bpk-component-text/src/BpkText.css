@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-text{margin:0}.bpk-text--xs{font-size:.75rem;line-height:1.125rem;font-weight:400;letter-spacing:normal}.bpk-text--sm{font-size:.75rem;line-height:1.125rem;font-weight:400;letter-spacing:normal}.bpk-text--base{font-size:1rem;line-height:1.5rem;font-weight:400;letter-spacing:normal}.bpk-text--lg{font-size:1.5rem;line-height:1.875rem;font-weight:400;letter-spacing:normal}.bpk-text--xl{font-size:1.75rem;line-height:2.625rem;font-weight:400;letter-spacing:normal}.bpk-text--xxl{font-size:2.625rem;line-height:3.375rem;font-weight:400;letter-spacing:normal}.bpk-text--bold{font-weight:700}
