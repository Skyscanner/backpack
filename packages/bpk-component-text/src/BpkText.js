/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';

import { cssModules, deprecated } from '../../bpk-react-utils';

import STYLES from './BpkText.module.scss';

const getClassName = cssModules(STYLES);

export const TEXT_STYLES = {
  xs: 'xs',
  sm: 'sm',
  base: 'base',
  lg: 'lg',
  xl: 'xl',
  xxl: 'xxl',
  xxxl: 'xxxl',
  xxxxl: 'xxxxl',
  xxxxxl: 'xxxxxl',
  caption: 'caption',
  footnote: 'footnote',
  label1: 'label-1',
  label2: 'label-2',
  label3: 'label-3',
  bodyDefault: 'body-default',
  bodyLongform: 'body-longform',
  subheading: 'subheading',
  heading1: 'heading-1',
  heading2: 'heading-2',
  heading3: 'heading-3',
  heading4: 'heading-4',
  heading5: 'heading-5',
  hero1: 'hero-1',
  hero2: 'hero-2',
  hero3: 'hero-3',
  hero4: 'hero-4',
  hero5: 'hero-5',
};

export const WEIGHT_STYLES = {
  regular: 'regular',
  bold: 'bold',
  black: 'black',
};

export type Weight = $Keys<typeof WEIGHT_STYLES>;
export type TextStyle = $Values<typeof TEXT_STYLES>;

const getWeight = (bold, weight, textStyle) => {
  if (bold || weight === WEIGHT_STYLES.bold) {
    return WEIGHT_STYLES.bold;
  }

  // Weight can only be black if textStyle is `xl`, `xxl` or `xxxl`.
  if (weight === WEIGHT_STYLES.black && textStyle.match(/^x+l$/)) {
    return WEIGHT_STYLES.black;
  }

  return null;
};

type Props = {
  children: Node,
  textStyle: TextStyle,
  tagName: 'span' | 'p' | 'text' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6',
  className: ?string,
  bold: ?boolean,
  weight: ?Weight,
};

const BpkText = (props: Props) => {
  const {
    bold,
    children,
    className,
    tagName: TagName,
    textStyle,
    weight,
    ...rest
  } = props;

  const validWeight = getWeight(bold, weight, textStyle);

  const classNames = getClassName(
    'bpk-text',
    `bpk-text--${textStyle}`,
    validWeight && `bpk-text--${validWeight}`,
    className,
  );

  return (
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    <TagName className={classNames} {...rest}>
      {children}
    </TagName>
  );
};

BpkText.propTypes = {
  children: PropTypes.node.isRequired,
  textStyle: PropTypes.oneOf(Object.values(TEXT_STYLES)),
  tagName: PropTypes.oneOf([
    'span',
    'p',
    'text',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
  ]),
  className: PropTypes.string,
  bold: deprecated(
    PropTypes.bool,
    'Use a different "textStyle" to achieve the desired weight.',
  ),
  weight: deprecated(
    PropTypes.oneOf(Object.keys(WEIGHT_STYLES)),
    'Use a different "textStyle" to achieve the desired weight.',
  ),
};

BpkText.defaultProps = {
  textStyle: TEXT_STYLES.bodyDefault,
  tagName: 'span',
  className: null,
  bold: null,
  weight: null,
};

export default BpkText;
