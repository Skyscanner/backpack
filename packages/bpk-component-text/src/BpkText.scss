@import '~bpk-mixins';

.bpk-text {
  @include bpk-text;

  &--xs {
    @include bpk-text--xs;
  }

  &--sm {
    @include bpk-text--sm;
  }

  &--base {
    @include bpk-text--base;
  }

  &--lg {
    @include bpk-text--lg;
  }

  &--xl {
    @include bpk-text--xl;
  }

  &--xxl {
    @include bpk-text--xxl;
  }

  &--xxxl {
    @include bpk-text--xxxl;
  }

  &--xxxxl {
    @include bpk-text--xxxxl;
  }

  &--xxxxxl {
    @include bpk-text--xxxxxl;
  }

  &--bold {
    @include bpk-text--bold;
  }

  &--black {
    @include bpk-text--black;
  }
}
