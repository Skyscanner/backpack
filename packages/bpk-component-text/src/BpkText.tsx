/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ReactNode } from 'react';

import { cssModules } from '../../bpk-react-utils';

import STYLES from './BpkText.module.scss';

const getClassName = cssModules(STYLES);

export const TEXT_STYLES = {
  xs: 'xs',
  sm: 'sm',
  base: 'base',
  lg: 'lg',
  xl: 'xl',
  xxl: 'xxl',
  xxxl: 'xxxl',
  xxxxl: 'xxxxl',
  xxxxxl: 'xxxxxl',
  caption: 'caption',
  footnote: 'footnote',
  label1: 'label-1',
  label2: 'label-2',
  label3: 'label-3',
  bodyDefault: 'body-default',
  bodyLongform: 'body-longform',
  subheading: 'subheading',
  heading1: 'heading-1',
  heading2: 'heading-2',
  heading3: 'heading-3',
  heading4: 'heading-4',
  heading5: 'heading-5',
  hero1: 'hero-1',
  hero2: 'hero-2',
  hero3: 'hero-3',
  hero4: 'hero-4',
  hero5: 'hero-5',
} as const;

export type TextStyle = (typeof TEXT_STYLES)[keyof typeof TEXT_STYLES];
export type Tag =
  | 'span'
  | 'p'
  | 'text'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6';

type Props = {
  children: ReactNode;
  textStyle?: TextStyle;
  tagName?: Tag;
  className?: string | null;
  id?: string;
  [rest: string]: any;
};

const BpkText = ({
  children,
  className = null,
  tagName: TagName = 'span',
  textStyle = TEXT_STYLES.bodyDefault,
  ...rest
}: Props) => {
  const classNames = getClassName(
    'bpk-text',
    `bpk-text--${textStyle}`,
    className,
  );

  return (
    <TagName className={classNames} {...rest}>
      {children}
    </TagName>
  );
};

export default BpkText;
