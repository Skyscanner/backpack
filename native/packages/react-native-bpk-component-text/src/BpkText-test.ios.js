/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2018 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkText from './BpkText';
import commonTests from './BpkText-test.common';

describe('iOS', () => {
  commonTests();

  it('should not apply `emphasize` for textStyle="xxl"', () => {
    jest.spyOn(console, 'error').mockImplementation(() => null);
    const tree = renderer
      .create(
        <BpkText emphasize textStyle="xxl">
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
          commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
          et magnis dis parturient montes, nascetur ridiculus mus.
        </BpkText>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
