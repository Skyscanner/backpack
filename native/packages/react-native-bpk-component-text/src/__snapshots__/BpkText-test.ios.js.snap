// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`iOS BpkText should render correctly 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 15,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with \`emphasize\` prop 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 15,
        "fontWeight": "400",
      },
      Object {
        "fontWeight": "600",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="base" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 15,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="lg" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 17,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="sm" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 13,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="xl" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 20,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="xs" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 11,
        "fontWeight": "400",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should render correctly with textStyle="xxl" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 34,
        "fontWeight": "700",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS BpkText should support overwriting styles 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 15,
        "fontWeight": "400",
      },
      Object {
        "color": "rgb(255, 84, 82)",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;

exports[`iOS should not apply \`emphasize\` for textStyle="xxl" 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  style={
    Array [
      Object {
        "color": "rgb(82, 76, 97)",
        "fontFamily": "System",
        "fontSize": 34,
        "fontWeight": "700",
      },
    ]
  }
>
  Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
</Text>
`;
